File: src/lib/verify/shell.ts
import { homeDir, listDir, shellHistory, type DirEntry, type HistoryFile } from "~/lib/shell";
import { fail, note, pass, type CheckResult, type Verifier } from "./types";

/**
 * The five terminal and navigation challenges.
 *
 * The Git verifiers have it easy: `git status` reports the result of the work.
 * `pwd`, `ls` and `cd` report nothing and change nothing, so checking them
 * needs a different substrate — the shell's own history file. That is what
 * keeps these real challenges rather than "mark as read" buttons, and it also
 * closes the obvious hole in the two filesystem ones, where a learner could
 * make the folders in Explorer and never open a terminal.
 *
 * Reading history is disclosed in the challenge text, and every result that
 * depends on it names the file it came from.
 */

// --- reading a command line ------------------------------------------------

export interface Invocation {
  name: string;
  args: string[];
}

/**
 * The commands in one history line.
 *
 * A line is often several: `mkdir notes && cd notes` has to count as both a
 * `mkdir` and a `cd`, or a learner who chains — which the challenges
 * themselves encourage by the end — fails a check they actually completed.
 */
export function invocations(line: string): Invocation[] {
  const parts: Invocation[] = [];

  for (const segment of line.split(/\s*(?:&&|\|\||[;|])\s*/)) {
    const words = segment.match(/"[^"]*"|'[^']*'|\S+/g);
    if (!words || words.length === 0) continue;

    // `VAR=x cmd` and `sudo cmd` both put the real command later in the line.
    let at = 0;
    while (at < words.length && (/^[A-Za-z_][A-Za-z0-9_]*=/.test(words[at]) || words[at] === "sudo")) {
      at++;
    }
    if (at >= words.length) continue;

    parts.push({ name: words[at], args: words.slice(at + 1) });
  }

  return parts;
}

/** Every command in every history file, most recent last. */
function allInvocations(files: HistoryFile[]): Invocation[] {
  return files.flatMap((file) => file.commands.flatMap(invocations));
}

/** Did any recorded command named `name` satisfy `where`? */
function ran(
  files: HistoryFile[],
  name: string,
  where: (inv: Invocation) => boolean = () => true,
): boolean {
  return allInvocations(files).some((inv) => inv.name === name && where(inv));
}

// --- history plumbing ------------------------------------------------------

/**
 * Bash, zsh and fish only write history when the shell exits, so a learner who
 * leaves the terminal open sees every check fail while having done the work.
 * That is the single most likely way these challenges waste someone's time, so
 * the fix is spelled out in the failure itself rather than left to a tip box.
 */
function flushHint(files: HistoryFile[]): string {
  const deferred = files.some((f) => f.shell !== "powershell");
  return deferred
    ? " If your terminal is still open, your shell has not saved its history yet — type `exit` to close it, open a new one, then press Check again."
    : "";
}

const NO_HISTORY = fail(
  "Could not find a shell history file. Git Gud reads Bash (~/.bash_history), " +
    "Zsh, Fish and PowerShell history. If you are on Windows using cmd.exe, switch to " +
    "Git Bash or PowerShell — cmd.exe keeps no history. Otherwise, run a command, close " +
    "the terminal so your shell saves its history, and check again.",
);

/** One line naming what was read, so nothing about this is hidden. */
function source(files: HistoryFile[]): CheckResult {
  const names = files.map((f) => `${f.shell} (${f.path})`).join(", ");
  return pass(`Read your shell history — ${names}`);
}

/** `ran`, wrapped as a result with the flush hint on failure. */
function didRun(
  files: HistoryFile[],
  label: string,
  name: string,
  where?: (inv: Invocation) => boolean,
): CheckResult {
  return ran(files, name, where)
    ? pass(`You ran ${label}.`)
    : fail(`No ${label} in your recent history.${flushHint(files)}`);
}

// --- paths -----------------------------------------------------------------

const isAbsolute = (arg: string) => /^([/~]|[A-Za-z]:)/.test(arg);

/** Compare two paths for "same place", across separators and Windows casing. */
function samePath(a: string, b: string): boolean {
  const normalize = (p: string) => p.replace(/\\/g, "/").replace(/\/+$/, "").toLowerCase();
  return normalize(a) === normalize(b);
}

/** Entries of `path`, or `null` when it does not exist. */
async function entriesOf(path: string): Promise<DirEntry[] | null> {
  try {
    return await listDir(path);
  } catch {
    return null;
  }
}

const hasFile = (entries: DirEntry[], name: string) =>
  entries.some((e) => !e.isDir && e.name.toLowerCase() === name.toLowerCase());

const hasDir = (entries: DirEntry[], name: string) =>
  entries.some((e) => e.isDir && e.name.toLowerCase() === name.toLowerCase());

/** Last segment of a path, separator-agnostic. */
function baseName(path: string): string {
  return path.replace(/[\\/]+$/, "").split(/[\\/]/).pop() ?? "";
}

// --- 1. Meet the Terminal --------------------------------------------------

/**
 * Proof that a terminal was opened and typed into. `whoami` and `date` are the
 * two safest commands on every platform — both exist in Git Bash, macOS, Linux
 * and PowerShell, and neither can damage anything if mistyped.
 */
export const verifyMeetTheTerminal: Verifier = async () => {
  const files = await shellHistory();
  if (files.length === 0) return [NO_HISTORY];

  return [
    source(files),
    didRun(files, "`whoami`", "whoami"),
    didRun(files, "`date`", "date"),
    // Clicking the window's close button teaches the same thing and leaves no
    // trace, so this can encourage `exit` but must never fail anyone for it.
    note("Closed the terminal with `exit`", ran(files, "exit")),
  ];
};

// --- 2. Command Performance ------------------------------------------------

/**
 * Command, arguments, options — checked as three separate shapes rather than
 * three specific commands, because the lesson is the grammar, not the verbs.
 */
export const verifyCommandPerformance: Verifier = async () => {
  const files = await shellHistory();
  if (files.length === 0) return [NO_HISTORY];

  const usedHelp = allInvocations(files).some((inv) =>
    inv.args.some((a) => a === "--help" || a === "-h" || a === "/?"),
  );

  return [
    source(files),
    didRun(
      files,
      "`echo` with something to say",
      "echo",
      (inv) => inv.args.length > 0,
    ),
    didRun(
      files,
      "`ls` with an option (like `ls -a` or `ls -l`)",
      "ls",
      (inv) => inv.args.some((a) => /^-[A-Za-z]/.test(a)),
    ),
    usedHelp
      ? pass("You asked a command to explain itself with `--help`.")
      : fail(`No \`--help\` in your recent history.${flushHint(files)}`),
    note("Tidied up with `clear`", ran(files, "clear") || ran(files, "cls")),
  ];
};

// --- 3. You Are Here -------------------------------------------------------

/**
 * `pwd` and `ls` leave no trace, so history carries them. The third check is
 * the interesting one: the learner points the folder picker at what they
 * believe is their home directory, and the app compares it against the home
 * directory it resolves independently. That turns "home directory and `~`"
 * from a definition into something with a right answer.
 */
export const verifyYouAreHere: Verifier = async (ctx) => {
  const files = await shellHistory();
  const results: CheckResult[] = [];

  if (files.length === 0) {
    results.push(NO_HISTORY);
  } else {
    results.push(
      source(files),
      didRun(files, "`pwd`", "pwd"),
      didRun(files, "`ls`", "ls"),
    );
  }

  const home = await homeDir();
  if (!home) {
    results.push(note("Could not work out your home directory on this system", false));
    return results;
  }

  if (!ctx.path) {
    results.push(fail(`Pick a folder above — the one \`~\` stands for on your computer.`));
  } else if (samePath(ctx.path, home)) {
    results.push(
      pass(`You picked ${ctx.path} — that is your home directory, the one \`~\` is short for.`),
    );
  } else {
    results.push(
      fail(
        `You picked ${ctx.path}, which is not your home directory. ` +
          `Open a terminal, run \`pwd\` before going anywhere, and pick the folder it prints.`,
      ),
    );
  }

  return results;
};

// --- 4. There and Back Again -----------------------------------------------

/**
 * Four shapes of `cd`, one per idea in the lesson: down into a child by a
 * relative name, up with `..`, home with `~`, and anywhere at all with an
 * absolute path. Checking the shapes rather than specific destinations means
 * the challenge works from whatever folders the learner actually has.
 */
export const verifyThereAndBackAgain: Verifier = async () => {
  const files = await shellHistory();
  if (files.length === 0) return [NO_HISTORY];

  const cdArg = (where: (arg: string) => boolean) => (inv: Invocation) =>
    inv.args.filter((a) => !a.startsWith("-")).some(where);

  return [
    source(files),
    didRun(
      files,
      "`cd` into a folder by its name (a relative path)",
      "cd",
      cdArg((a) => !isAbsolute(a) && !a.startsWith("..")),
    ),
    didRun(files, "`cd ..` to go up to the parent folder", "cd", cdArg((a) => a.startsWith(".."))),
    didRun(
      files,
      "`cd ~` to jump home",
      "cd",
      (inv) => inv.args.length === 0 || inv.args.includes("~"),
    ),
    didRun(
      files,
      "`cd` with an absolute path",
      "cd",
      cdArg((a) => isAbsolute(a) && a !== "~"),
    ),
    note("Checked where you landed with `pwd`", ran(files, "pwd")),
  ];
};

// --- 5. Make It So ---------------------------------------------------------

/** The tree challenge 5 asks for, relative to the folder the learner picks. */
export const PRACTICE_DIR = "gitgud-practice";

export const verifyMakeItSo: Verifier = async (ctx) => {
  const results: CheckResult[] = [];

  if (!ctx.path) {
    return [fail(`Pick your \`${PRACTICE_DIR}\` folder above, then check again.`)];
  }

  results.push(
    samePath(baseName(ctx.path), PRACTICE_DIR)
      ? pass(`Found \`${PRACTICE_DIR}\`.`)
      : fail(
          `You picked \`${baseName(ctx.path)}\`, not \`${PRACTICE_DIR}\`. ` +
            `Pick the folder you made with \`mkdir ${PRACTICE_DIR}\`.`,
        ),
  );

  const root = await entriesOf(ctx.path);
  if (!root) {
    results.push(fail(`Could not read ${ctx.path}. Has it been moved or deleted?`));
    return results;
  }

  results.push(
    hasFile(root, "hello.txt")
      ? pass("`hello.txt` is there.")
      : fail("No `hello.txt` inside. Run `touch hello.txt` in that folder."),
  );

  // Checked level by level so the failure names the step that is missing,
  // rather than reporting "notes/day-one.txt" for someone who never made
  // `notes` at all.
  const notes = hasDir(root, "notes") ? await entriesOf(`${ctx.path}/notes`) : null;

  if (!notes) {
    results.push(fail("No `notes` folder inside. Run `mkdir notes`."));
  } else {
    results.push(
      hasFile(notes, "day-one.txt")
        ? pass("`notes/day-one.txt` is there.")
        : fail("`notes` is there but empty. Run `touch notes/day-one.txt`."),
    );

    const archive = hasDir(notes, "archive") ? await entriesOf(`${ctx.path}/notes/archive`) : null;
    results.push(
      archive && hasDir(archive, "2026")
        ? pass("`notes/archive/2026` is there — `mkdir -p` built all three at once.")
        : fail("No `notes/archive/2026`. Run `mkdir -p notes/archive/2026`."),
    );
  }

  // Without this, making the folders in Explorer would pass. The whole point
  // of the module is that the terminal did it.
  const files = await shellHistory();
  if (files.length === 0) {
    results.push(NO_HISTORY);
  } else {
    results.push(
      source(files),
      didRun(files, "`mkdir` in your terminal", "mkdir"),
      didRun(files, "`touch` in your terminal", "touch"),
    );
  }

  return results;
};

File: src-tauri/src/shell.rs
//! Filesystem and shell-history probes, for the terminal challenges.
//!
//! The Git challenges can check their work by asking Git. The terminal ones
//! cannot: `pwd`, `ls` and `cd` change nothing on disk, so there is no state
//! to inspect afterwards. Two probes cover the gap.
//!
//! `list_dir` checks the *result* of `mkdir` and `touch`.
//!
//! `history` checks the *act* of running a command, by reading the history
//! file the user's own shell keeps. That is what makes "run `pwd`" a real
//! challenge rather than a checkbox — and it also closes the hole where a
//! learner creates the folders in Explorer and never opens a terminal at all.
//!
//! Only the last few hundred commands are returned, they never leave the
//! machine, and every challenge that reads them names the file in its results.
//! The challenge text says so before the button is pressed.

use std::fs;
use std::path::{Path, PathBuf};

use serde::Serialize;

/// How far back a verifier can see. Enough for a long session, short enough
/// that an unrelated `pwd` from last month cannot pass a challenge.
const MAX_COMMANDS: usize = 300;

#[derive(Debug, Clone, Serialize)]
pub struct Entry {
    pub name: String,
    pub is_dir: bool,
}

/// Immediate children of `path`. Not recursive — a verifier that needs a
/// nested path calls again, which keeps this from ever walking a home
/// directory.
pub fn list_dir(path: &str) -> Result<Vec<Entry>, String> {
    let read = fs::read_dir(path).map_err(|e| match e.kind() {
        std::io::ErrorKind::NotFound => format!("That folder does not exist: {path}"),
        std::io::ErrorKind::PermissionDenied => format!("No permission to read {path}"),
        _ => format!("Could not read {path}: {e}"),
    })?;

    let mut entries = Vec::new();
    for item in read.flatten() {
        // `file_type()` rather than `metadata()`: it does not follow symlinks,
        // so a link pointing at a directory is not reported as one.
        let is_dir = item.file_type().map(|t| t.is_dir()).unwrap_or(false);
        entries.push(Entry {
            name: item.file_name().to_string_lossy().into_owned(),
            is_dir,
        });
    }

    Ok(entries)
}

#[derive(Debug, Clone, Serialize)]
pub struct HistoryFile {
    /// `bash`, `zsh`, `fish` or `powershell` — named in the verify results so
    /// the learner knows which file was read.
    pub shell: String,
    pub path: String,
    pub commands: Vec<String>,
}

/// Every shell history file we can find under `home`.
///
/// An empty list means no supported shell has written one yet, which is a
/// different answer from "the command was not run" — the verifiers say so.
pub fn history(home: &Path) -> Vec<HistoryFile> {
    let mut found = Vec::new();

    for (shell, path) in candidates(home) {
        let Ok(bytes) = fs::read(&path) else { continue };
        // A history file is in whatever encoding the terminal used. A stray
        // non-UTF-8 byte must not cost the learner the whole file.
        let text = String::from_utf8_lossy(&bytes);
        let commands = parse(&shell, &text);
        if commands.is_empty() {
            continue;
        }
        found.push(HistoryFile {
            shell,
            path: path.to_string_lossy().into_owned(),
            commands,
        });
    }

    found
}

fn candidates(home: &Path) -> Vec<(String, PathBuf)> {
    let mut list = vec![
        ("bash".to_string(), home.join(".bash_history")),
        ("zsh".to_string(), home.join(".zsh_history")),
        (
            "fish".to_string(),
            home.join(".local/share/fish/fish_history"),
        ),
        (
            "powershell".to_string(),
            home.join(".local/share/powershell/PSReadLine/ConsoleHost_history.txt"),
        ),
    ];

    // PSReadLine on Windows writes under APPDATA, which is not below the home
    // directory on a roaming profile, so it has to be resolved separately.
    if let Some(appdata) = std::env::var_os("APPDATA") {
        list.push((
            "powershell".to_string(),
            PathBuf::from(appdata)
                .join("Microsoft/Windows/PowerShell/PSReadLine/ConsoleHost_history.txt"),
        ));
    }

    list
}

/// One command per entry, in the file's own order, most recent last.
///
/// Each shell writes its own dialect, and all of the non-bash ones would
/// otherwise come back as unmatchable noise: zsh prefixes `: <epoch>:<n>;`,
/// fish writes a YAML-ish record, and bash writes bare `#<epoch>` lines once
/// `HISTTIMEFORMAT` is set.
fn parse(shell: &str, text: &str) -> Vec<String> {
    let mut commands: Vec<String> = Vec::new();

    for line in text.lines() {
        let line = line.trim_end_matches('\r');

        let command = match shell {
            "fish" => match line.strip_prefix("- cmd: ") {
                Some(rest) => rest,
                None => continue,
            },
            "zsh" => match line.strip_prefix(": ") {
                // `: 1700000000:0;ls -la` — the command follows the first `;`.
                Some(rest) => match rest.split_once(';') {
                    Some((_, command)) => command,
                    None => continue,
                },
                None => line,
            },
            // A bare `#1700000000` is bash's timestamp record, not a command.
            _ => {
                if line.starts_with('#') {
                    continue;
                }
                line
            }
        };

        let command = command.trim();
        if !command.is_empty() {
            commands.push(command.to_string());
        }
    }

    if commands.len() > MAX_COMMANDS {
        commands.drain(..commands.len() - MAX_COMMANDS);
    }
    commands
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn zsh_extended_history_is_unwrapped() {
        let text = ": 1700000000:0;ls -la\n: 1700000001:0;pwd\n";
        assert_eq!(parse("zsh", text), vec!["ls -la", "pwd"]);
    }

    #[test]
    fn plain_zsh_history_still_parses() {
        // `EXTENDED_HISTORY` is off by default, so the same file may hold bare
        // lines. Treating those as unparseable would fail every macOS learner
        // who never set the option.
        assert_eq!(parse("zsh", "pwd\nls\n"), vec!["pwd", "ls"]);
    }

    #[test]
    fn bash_timestamp_records_are_not_commands() {
        let text = "#1700000000\npwd\n#1700000001\nls -a\n";
        assert_eq!(parse("bash", text), vec!["pwd", "ls -a"]);
    }

    #[test]
    fn fish_records_yield_only_the_command() {
        let text = "- cmd: pwd\n  when: 1700000000\n- cmd: ls\n  when: 1700000001\n";
        assert_eq!(parse("fish", text), vec!["pwd", "ls"]);
    }

    #[test]
    fn only_the_most_recent_commands_are_returned() {
        // Without the cap, a `pwd` typed months ago would pass challenge 3
        // forever — the same "passes without doing the work" failure the Git
        // verifiers had to be fixed for.
        let text = (0..MAX_COMMANDS + 50)
            .map(|i| format!("echo {i}"))
            .collect::<Vec<_>>()
            .join("\n");
        let parsed = parse("bash", &text);
        assert_eq!(parsed.len(), MAX_COMMANDS);
        assert_eq!(
            parsed.last().unwrap(),
            &format!("echo {}", MAX_COMMANDS + 49)
        );
    }

    #[test]
    fn history_finds_and_labels_each_shells_file() {
        // The parsing is covered above; this covers discovery — the paths and
        // the shell labels, which are what the verify results quote back to
        // the learner. A typo in one of those paths would silently make every
        // terminal challenge unpassable on that platform.
        let home = std::env::temp_dir().join("gitgud-history-test");
        let _ = fs::remove_dir_all(&home);
        fs::create_dir_all(home.join(".local/share/fish")).unwrap();

        fs::write(home.join(".bash_history"), "pwd\nls -a\n").unwrap();
        fs::write(home.join(".zsh_history"), ": 1700000000:0;whoami\n").unwrap();
        fs::write(
            home.join(".local/share/fish/fish_history"),
            "- cmd: date\n  when: 1700000000\n",
        )
        .unwrap();

        let found = history(&home);
        let by_shell = |name: &str| found.iter().find(|f| f.shell == name);

        assert_eq!(by_shell("bash").unwrap().commands, vec!["pwd", "ls -a"]);
        assert_eq!(by_shell("zsh").unwrap().commands, vec!["whoami"]);
        assert_eq!(by_shell("fish").unwrap().commands, vec!["date"]);
        // The path is shown to the learner, so it has to be the real one.
        assert!(by_shell("bash").unwrap().path.ends_with(".bash_history"));

        let _ = fs::remove_dir_all(&home);
    }

    #[test]
    fn an_empty_history_file_is_not_reported_as_a_source() {
        // "Read your bash history" followed by every check failing is a worse
        // answer than "no history found", which at least names the fix.
        let home = std::env::temp_dir().join("gitgud-empty-history-test");
        let _ = fs::remove_dir_all(&home);
        fs::create_dir_all(&home).unwrap();
        fs::write(home.join(".bash_history"), "\n\n").unwrap();

        assert!(history(&home).iter().all(|f| f.shell != "bash"));

        let _ = fs::remove_dir_all(&home);
    }

    #[test]
    fn list_dir_separates_files_from_directories() {
        let root = std::env::temp_dir().join("gitgud-listdir-test");
        let _ = fs::remove_dir_all(&root);
        fs::create_dir_all(root.join("notes")).unwrap();
        fs::write(root.join("hello.txt"), "").unwrap();

        let mut entries = list_dir(root.to_str().unwrap()).unwrap();
        entries.sort_by(|a, b| a.name.cmp(&b.name));

        assert_eq!(entries.len(), 2);
        assert_eq!(entries[0].name, "hello.txt");
        assert!(!entries[0].is_dir);
        assert_eq!(entries[1].name, "notes");
        assert!(entries[1].is_dir);

        let _ = fs::remove_dir_all(&root);
    }

    #[test]
    fn a_missing_folder_is_an_error_not_an_empty_listing() {
        // Empty and missing must stay distinguishable: "your folder is empty"
        // and "you never made that folder" are different lessons.
        let missing = std::env::temp_dir().join("gitgud-does-not-exist-xyz");
        let _ = fs::remove_dir_all(&missing);
        assert!(list_dir(missing.to_str().unwrap()).is_err());
    }
}
